### PR TITLE
Hunter and Warrior Balance Tweak

### PR DESCRIPTION
**Pull request by Surrealaser**
:cl: by Surrealistik
tweak: RNG on grab breaking removed; now whenever you resist, you increment your grab resist level by 1 until it exceeds the level of a grab, thereby breaking it.
tweak: Hunters gain armor penetration on stationary/stalk sneak attacks: 20% while on harm intent, and 50% while on disarm intent.
tweak: Stagger, slowdown stacks decreased by 33% for running sneak attacks. Stagger and slowdown stacks increased by 33% for stationary/stalk sneak attacks. Knockout stacks reduced by 50% for running sneak attacks.
tweak: Hunter sneak attack multiplier now declines by 0.2 per tile moved while running to a minimum of 1.25x. This is recovered at a rate of +1 per tick while stationary/destealthed, or per tile moved while walking/stalking.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/living.dm [hunk 1]
```diff
@@ -277,17 +277,19 @@
...
-		if(prob(30/pulledby.grab_level))
+		grab_resist_level += 1
+		if(grab_resist_level > pulledby.grab_level)
 			playsound(src.loc, 'sound/weapons/thudswoosh.ogg', 25, 1, 7)
 			visible_message("<span class='danger'>[src] has broken free of [pulledby]'s grip!</span>", null, null, 5)
 			pulledby.stop_pulling()
-			return 1
+			grab_resist_level = 0 //zero it out.
+			return TRUE
 		if(moving_resist && client) //we resisted by trying to move
```
Comment (DominikPanic) [inline code/modules/mob/living/living.dm]: Use the proper changeNextMove proc now that the click rework is merged here

File: code/modules/mob/living/carbon/xenomorph/attack_alien.dm [hunk 2]
```diff
@@ -173,17 +174,26 @@
 
 			if(M.stealth_router(HANDLE_STEALTH_CHECK)) //Cancel stealth if we have it due to aggro.
 				if(M.stealth_router(HANDLE_SNEAK_ATTACK_CHECK)) //Pouncing prevents us from making a sneak attack for 4 seconds
-					if(m_intent == MOVE_INTENT_RUN)
+					#if DEBUG_ATTACK_ALIEN
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/attack_alien.dm]:
```suggestion
					#ifdef DEBUG_ATTACK_ALIEN
```

File: code/modules/mob/living/carbon/xenomorph/attack_alien.dm [hunk 3]
```diff
@@ -7,6 +7,7 @@
  * In that case, the first argument is always the attacker. For attack_alien, it should always be Xenomorph sub-types
  */
 
+#define DEBUG_ATTACK_ALIEN	0
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/attack_alien.dm]:
```suggestion
//#define DEBUG_ATTACK_ALIEN
```

File: code/modules/mob/living/carbon/xenomorph/attack_alien.dm [hunk 4]
```diff
@@ -224,30 +234,41 @@
 				tackle_pain = tackle_pain * (1 - (0.10 + 0.05 * protection_aura))  //Halloss damage decreased by 10% + 5% per rank of protection aura
 			if(M.stealth_router(HANDLE_STEALTH_CHECK))
 				if(M.stealth_router(HANDLE_SNEAK_ATTACK_CHECK))
-					KnockOut(2)
-					if(m_intent == MOVE_INTENT_RUN && ( last_move_intent > (world.time - 20) ) ) //Allows us to slash while running... but only if we've been stationary for awhile
+					#if DEBUG_ATTACK_ALIEN
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/attack_alien.dm]:
```suggestion
					#ifdef DEBUG_ATTACK_ALIEN
```

File: code/modules/mob/living/carbon/xenomorph/attack_alien.dm [hunk 5]
```diff
@@ -224,30 +234,41 @@
...
 			if(dam_bonus)
 				tackle_pain += dam_bonus
 
 			tackle_pain = M.hit_and_run_bonus(tackle_pain) //Apply Runner hit and run bonus damage if applicable
 
-			apply_damage(tackle_pain, HALLOSS, "chest", armor_block * 0.4) //Only half armour applies vs tackle
+			var/final_damage = apply_damage(tackle_pain, HALLOSS, "chest", armor_block * XENO_TACKLE_ARMOR_PEN)
 			updatehealth()
 			updateshock()
+			#if DEBUG_ATTACK_ALIEN
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/attack_alien.dm]:
```suggestion
			#ifdef DEBUG_ATTACK_ALIEN
```

File: code/modules/mob/living/carbon/human/human_damage.dm [hunk 6]
```diff
@@ -400,7 +400,7 @@ This function restores all limbs.
 	else
 		if(!def_zone)	def_zone = ran_zone(def_zone)
 		organ = get_limb(check_zone(def_zone))
-	if(!organ)	return 0
+	if(!organ)	return FALSE
```
Comment (DominikPanic) [inline code/modules/mob/living/carbon/human/human_damage.dm]: Same line control statement.

File: code/modules/mob/living/living.dm (hunk 1 above)
Comment (Surrealaser) [reply code/modules/mob/living/living.dm]: I don't know what you're talking about.

Comment (DominikPanic) [reply code/modules/mob/living/living.dm]: clinet.next_movement shouldn't be used, instead do something like
`usr.changeNext_move(value)`

File: code/__DEFINES/mobs.dm [hunk 7]
```diff
@@ -440,8 +440,30 @@ var/list/global_mutations = list() // list of hidden mutation things
 
 #define ANTI_CHAINSTUN_TICKS	2
 
-//Xeno Defines
+//Hunter Defines
+#define HUNTER_STEALTH_COOLDOWN					50 //5 seconds
```
Comment (Rohesie) [inline code/__DEFINES/mobs.dm]: Repeated vars. Merge garbage.

Comment (Surrealaser) [reply code/__DEFINES/mobs.dm]: What are you talking about?

**Comment by Surrealaser** [comment]
Can this fucking thing stop conflicting for one goddamn day

File: code/modules/mob/living/carbon/xenomorph/Castes/Hunter.dm [hunk 8]
```diff
@@ -0,0 +1,174 @@
+/datum/xeno_caste/hunter
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/Castes/Hunter.dm]: this has been rebased incorrectly, caste info is in the hunter/castedatum_hunter.dm file

File: code/modules/mob/living/carbon/xenomorph/XenoProcs.dm [hunk 9]
```diff
@@ -26,7 +26,7 @@
 
 		var/area/A = get_area(X)
 		xenoinfo += " <b><font color=green>([A ? A.name : null])</b></td></tr>"
-	
+
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/XenoProcs.dm]: this file has been renamed

Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/XenoProcs.dm]: It appears as lower case xenoprocs on my end; weird.

**Comment by spookydonut** [comment]
If no bugs after testing can be merged

File: code/__DEFINES/mobs.dm [hunk 10]
```diff
@@ -528,6 +528,17 @@ GLOBAL_LIST_INIT(xenoupgradetiers, list(XENO_UPGRADE_BASETYPE, XENO_UPGRADE_INVA
...
+#define HUNTER_SNEAK_SLASH_ARMOR_PEN			0.8 //1 - this value = the actual penetration
+#define HUNTER_SNEAK_ATTACK_RUN_DELAY			2 SECONDS
+#define HUNTER_SNEAKATTACK_MAX_MULTIPLIER		3.5
+#define HUNTER_SNEAKATTACK_RUN_REDUCTION		0.2
+#define HUNTER_SNEAKATTACK_WALK_INCREASE		1
+#define HUNTER_SNEAKATTACK_MULTI_RECOVER_DELAY	10
+
+// xeno defines
+
+#define XENO_TACKLE_ARMOR_PEN	0.4 //Actual armor pen is 1 - this value.
```
Comment (Rohesie) [inline code/__DEFINES/mobs.dm]: Where is this being used?

File: code/modules/mob/living/living_defines.dm [hunk 11]
```diff
@@ -88,4 +88,9 @@
 
 	var/ventcrawl_layer = PIPING_LAYER_DEFAULT
 
-	var/ventcrawl_message_busy
\ No newline at end of file
+	var/ventcrawl_message_busy
+
+	var/taken = FALSE
```
Comment (Rohesie) [inline code/modules/mob/living/living_defines.dm]: What is this var about?

File: code/modules/mob/living/living.dm [hunk 12]
```diff
@@ -278,17 +278,19 @@
 
 /mob/living/resist_grab(moving_resist)
 	if(pulledby.grab_level)
-		if(prob(30/pulledby.grab_level))
+		grab_resist_level += 1
```
Comment (Rohesie) [inline code/modules/mob/living/living.dm]: This doesn't get reset unless by breaking a grab. Means I can store grab resists to be used later in combat.

File: code/modules/mob/living/living_defines.dm (hunk 11 above)
Comment (DominikPanic) [reply code/modules/mob/living/living_defines.dm]: That is bad  conflict resolution I removed that var, it was for the offer mob

File: code/modules/mob/living/living.dm [hunk 13]
```diff
@@ -278,18 +278,26 @@
...
+		grab_resist_level = 0 //zero it out.
 		pulledby.stop_pulling()
-		return 1
+		return TRUE
 
+/mob/living/stop_pulling()
+	if(isliving(pulling))
+		var/mob/living/L = pulling
+		L.grab_resist_level = 0 //zero it out
+	..()
```
Comment (DominikPanic) [inline code/modules/mob/living/living.dm]:
```suggestion
	return ..()
```

File: code/modules/mob/living/living.dm (hunk 12 above)
Comment (ondrej008) [reply code/modules/mob/living/living.dm]: Looks like you're right.

1. Have someone get the strongest grab on you.
2. Resist until you're one resist away from breaking free.
3. Get grabbed by a xeno, resist once and you're free.
4. ???
5. Profit

Comment (Surrealaser) [reply code/modules/mob/living/living.dm]: This has been fixed by resets to grab_resist_level elsewhere

**Comment by Surrealaser** [comment]
@spookydonut Need to get this flagged/tested/then permamerged before we have another massive conflict to resolve

**Pull request merged**